File: solutions/02/2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func AbsInt(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	// open the file
	// https://gobyexample.com/reading-files
	file, err := os.Open("./solutions/02/input.txt")
	check(err)
	defer file.Close()

	// get a list of reports
	reports := [][]int{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, " ")
		ints := make([]int, len(parts))
		for i, s := range parts {
			num, err := strconv.Atoi(s)
			check(err)
			ints[i] = num
		}
		reports = append(reports, ints)
	}

	// score the reports
	numValid := 0
	for _, report := range reports {
		all_decreasing := true
		all_increasing := true
		all_diff_ok := true

		for i := 0; i < len(report)-1; i++ {
			prev, curr := report[i], report[i+1]
			all_decreasing = all_decreasing && (prev > curr)
			all_increasing = all_increasing && (prev < curr)
			all_diff_ok = all_diff_ok && (1 <= AbsInt(prev-curr) && AbsInt(prev-curr) <= 3)
		}

		safe := (all_increasing != all_decreasing) && all_diff_ok
		if safe {
			numValid += 1
		}

		fmt.Println(numValid)

	}

}

File: solutions/02/2.py
# https://adventofcode.com/2024/day/2
# open the input and build a lists of ints to process
input = "./solutions/02/input.txt"
# input = "./solutions/02/sample.txt"

with open(input, mode="r") as f:
    input = f.readlines()
    reports = [
        [int(x) for x in line.strip().split(" ")] 
        for line in input
    ]

num_valid = 0
for report in reports:
    all_increasing = True
    all_decreasing = True
    all_diff_ok = True

    for i in range(len(report) - 1):
        prev, curr = report[i], report[i + 1]

        all_decreasing = all_decreasing and (prev > curr)

        all_increasing = all_increasing and (prev < curr)

        all_diff_ok = all_diff_ok and (1 <= abs(prev - curr) <= 3)

    safe = (all_increasing ^ all_decreasing) and all_diff_ok

    if safe:
        num_valid += 1

print(f"valid reports: {num_valid}")












